Write lists of minions targeted by syndic masters to job cache

terminalmage opened the pull request:
## Summary

When a master-of-masters publishes a command, and a syndic master re-publishes it to its subscribed minions, it fires an event back to the master containing the list of minions matched by the syndic master for that job.

This pull request scans these events for lists of minions, and writes them to the job cache.

Fixes #30528.
## How to test
1. Use [this documentation](https://help.github.com/articles/checking-out-pull-requests-locally/) to check out this pull request into a new branch, and install from git.
2. Recommended setup: 3 hosts (1 master-of-masters, 1 syndic master, 1 lower-level minion). Set `order_masters: True` on the master-of-masters and subscribe the lower-level minion to the syndic master, and the syndic master to the master-of-masters.
3. From the master-of-masters, run `salt -v '*' test.ping`
4. Take the jid obtained in step 3 and run `salt-run jobs.list_job <jid>`. The `Minions` list in the return data should contain the minion ID of the lower-level minion. Prior to this pull request, the lower-level minions would not be in this return data. After completing the rest of the test steps, installing Salt from the head of `2015.8` on the master-of-masters and repeating steps 3 and 4 should show no lower-level minions in the `jobs.list_job` return data.
5. Stop the `salt-minion` service on the lower-level minion, and then repeat step 3.
6. Take the jid obtained in step 5 and run `salt-run jobs.lookup_jid <jid> missing=True`. The lower-level minion should be included in the return data, and should show `Minion did not return`. Repeating the call to `jobs.lookup_jid` without the `missing=True` should exclude the lower-level minion from the return data.
## Review notes

Of particular focus in the review of this pull request should be the following:
- [ ] The `couchbase_return` returner. To maintain API compatibility I needed to move the writing of minion lists into a separate function, as I did for the `local_cache` returner, and I did not have an instance to test this.
- [ ] The file locking code added in salt/utils/files.py.

terminalmage committed:
Add salt.utils.split_input()

terminalmage committed:
Add missing RST file for slsutil module

terminalmage committed:
Add a contextmanager for file locking

terminalmage committed:
Add an exception class for errors encountered while locking files.

terminalmage committed:
Simplify jobs.get_jobs logic, generally improve jobs runner docs

terminalmage committed:
Separate writing of serialized minion list into its own function

This allows minions to be updated when the master receives events
containing lists of minions targeted by lower-level masters.

terminalmage committed:
Add utils function to invoke a returner's save_minions() func

terminalmage committed:
Update a job's minion list to include minion lists forwarded by syndic

This allows the job cache to include minions matched by lower-level
masters for a given job.

terminalmage committed:
salt.returners.couchbase_return: move minion list updates to new save_minions() func

terminalmage committed:
salt.returners.postgres_local_cache: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.sqlite3_return: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.etcd_return: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.influxdb_return: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.cassandra_cql_return: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.mongo_return: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.pgjsonb: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.couchdb_return: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.mysql: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.memcache_return: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.postgres: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.odbc: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.mongo_future_return: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.multi_returner: add no-op save_minions() func for API compatibility

terminalmage committed:
salt.returners.redis_return: add no-op save_minions() func for API compatibility

jtand commented:
Go Go Jenkins!

jtand commented:
Go Go Jenkins!

jtand commented:
Go Go Jenkins!

terminalmage committed:
lint fixes

rallytime commented:
Go Go Jenkins!

cachedout commented on salt/returners/etcd_return.py:
```diff
@@ -113,6 +113,13 @@ def save_load(jid, load):
+    '''
+    Included for API consistency
+    '''
+    pass
```
I'm sort of inclined to raise exceptions for these and catch and log higher up.

terminalmage replied:
The only reason that I did not is that we're not already saving the minion list for this returner.

cachedout replied:
Fair enough. Good point.

cachedout commented:
(With apologies for being the person who brought up locking the first place.)

After looking at this, I'm wondering if instead of locking we consider splitting this write up into multiple files and then scanning those when we retrieve the minions list. (I'm talking about the local cache of course.) The locking approach is starting to worry me..

terminalmage committed:
Skip events with minion lists but no jid

Some events just contain a timestamp and a minion list, but no jid.
These can safely be ignored, lest they cause supurious log messages.

Also, pass the syndic's id so that it can be stored in a unique file in
the job cache (in the local_cache returner).

terminalmage committed:
Add syndic_id param for API compatibility

Also fix successive updates

terminalmage committed:
Add argument to save_minions() to pass a syndic ID

This is used to provide a unique filename for the minion list for a
given syndic.

terminalmage committed:
Pass syndic_id to save_minions()

terminalmage commented:
@cachedout Good idea, I just pushed some changes, let me know what you think.

cachedout commented:
Yeah, I think this is better. Let's go with this.
